### PR TITLE
Lazily initialize TransportActionStats in RequestHandlerRegistry

### PR DESCRIPTION
**Pull request by original-brownbear**
Same as #114107 but for the transport side.
We are consuming more than 0.5M for these stat instances and a large fraction if not the majority will go unused in many use-cases, given how expensive stats are in terms of volatile operations the overhead of an added `getAcquire` is trivial in the absence of writes.
Outside of reducing the prod footprint this is also kind of nice in x-pack internal cluster tests in particular where it saves quite a bit of heap.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java [hunk 1]
```diff
@@ -33,7 +35,19 @@ public class RequestHandlerRegistry<Request extends TransportRequest> implements
     private final TaskManager taskManager;
     private final Tracer tracer;
     private final Writeable.Reader<Request> requestReader;
-    private final TransportActionStatsTracker statsTracker = new TransportActionStatsTracker();
+    @SuppressWarnings("unused") // only accessed via #STATS_TRACKER_HANDLE, lazy initialized because instances consume non-trivial heap
+    private volatile TransportActionStatsTracker statsTracker;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java]: Right but lets drop the `volatile` then :D I figured I'd be fun to experiment with weaker ordering a little here and there in obvious spots, might be helpful on ARM in some cases where there's only ever a single write to a variable. Who knows might help init performance either now or on some future ARM or similar hardware.
(admittedly there's more interesting spots for that stuff than this ... :))

File: server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java [hunk 2]
```diff
@@ -33,7 +35,19 @@ public class RequestHandlerRegistry<Request extends TransportRequest> implements
...
+
+    private static final VarHandle STATS_TRACKER_HANDLE;
+
+    static {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java:43]: Sure thing, dded this one as well as `MethodHandlers` where I did the same thing now.

File: server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java [hunk 3]
```diff
@@ -118,15 +132,34 @@ public static <R extends TransportRequest> RequestHandlerRegistry<R> replaceHand
...
     public TransportActionStats getStats() {
+        var statsTracker = existingStatsTracker();
+        if (statsTracker == null) {
+            return TransportActionStats.EMPTY;
+        }
         return statsTracker.getStats();
     }
+
+    private TransportActionStatsTracker statsTracker() {
+        var tracker = existingStatsTracker();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java:152]: If you're looking for yet-weaker memory semantics, I think this one can be a regular (nonvolatile) read. If it's null then the `compareAndExchange` will establish the right guarantees, and if it's not null then it has to be the right value.

Risks perhaps creating a few more unused `TransportActionStatsTracker` instances tho I guess.

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/RequestHandlerRegistry.java:152]: Yea somehow a plain read in this kind of spot weirds me out a little :D but you are right, we could relax these spots further technically I guess I just don't feel adventurous enough for that this evening (I do realise though that we actually do this kind of thing ... sync write -> plain read ... in a couple spots like the CS + routing table computation already).

**Pull request merged**